WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,662 @@
+      pComrvCB->stOverlayCache[ucIndex].pFixedEntryAddress             = pBaseAddress;
+      pBaseAddress = ((u08_t*)pBaseAddress + D_COMRV_OVL_GROUP_SIZE_MIN);
+   }
+   /* mark the last entry in the LRU list */
+   pComrvCB->stOverlayCache[ucIndex-1].ucNextIndex = D_COMRV_LRU_LAST_ITEM;
+   /* set the index of the list LRU and MRU */
+   pComrvCB->ucLruIndex = 0;
+   pComrvCB->ucMruIndex = ucIndex-1;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+   pStackStartAddr = (comrvStackFrame_t*)&pComrvCB->stOverlayCache[pComrvCB->ucNumOfOverlayEntries];
+   /* Init comrv stack - mark the last stack frame */
+   pStackStartAddr->ssOffsetPrevFrame = D_COMRV_END_OF_STACK;
+   /* initialize all stack entries */
+   for (pStackStartAddr++ ; pStackStartAddr < pStackEndAddr ; pStackStartAddr++)
+   {
+      pStackStartAddr->ssOffsetPrevFrame = (s16_t)-sizeof(comrvStackFrame_t);
+      pStackStartAddr->uiCalleeToken = 0;
+   }
+
+#ifndef D_COMRV_USE_OS
+   /* set the address of COMRV stack and initialize it */
+   pStackStartAddr--;
+   M_COMRV_SET_STACK_ADDR(pStackStartAddr);
+   pStackStartAddr->ssOffsetPrevFrame = D_COMRV_END_OF_STACK;
+#endif /* D_COMRV_USE_OS */
+
+   /* set the address of COMRV stack pool */
+   pStackStartAddr--;
+   M_COMRV_SET_POOL_ADDR(pStackStartAddr);
+
+   /* set the address of COMRV entry point */
+   M_COMRV_SET_ENTRY_ADDR(comrvEntry);
+}
+
+/**
+* Search if current overlay token is already loaded
+*
+* @param none
+*
+* @return void* - address of the overlay function/data
+*/
+void* comrvGetAddressFromToken(void)
+{
+   comrvOverlayToken_t unToken;
+   comrvCacheEntry_t*   pEntry;
+   u08_t               ucIsInvoke;
+   void*               pAddress;
+   comrvStackFrame_t*  pComrvStackFrame;
+   u16_t               usOverlayGroupSize, usOffset;
+   u08_t               ucNumOfEvictionCandidates, ucIndex, ucSizeOfEvictionCandidates;
+   u08_t               ucEntryIndex, ucEvictCandidateList[D_COMRV_CANDIDATE_LIST_SIZE];
+#ifdef D_COMRV_CRC
+   u32_t               uiCrc;
+#endif /* D_COMRV_CRC */
+#ifdef D_COMRV_FW_PROFILING
+   u32_t               uiProfilingIndication;
+#endif /* D_COMRV_FW_PROFILING */
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+   u32_t               uiTemp;
+   u16_t               usSelectedMultiGroupEntry;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* read the requested token value */
+   M_COMRV_READ_TOKEN_REG(unToken.uiValue);
+
+   /* read comrv stack register (t3) */
+   M_COMRV_READ_STACK_REG(pComrvStackFrame)
+
+   /* get the invoke callee indication */
+   ucIsInvoke = (u32_t)pComrvStackFrame & D_COMRV_INVOKE_CALLEE_BIT_0;
+   /* are we calling a new overlay function */
+   if (ucIsInvoke == D_COMRV_INVOKE_CALLEE_BIT_0)
+   {
+      /* clear the invoke indication from pComrvStackFrame address */
+      pComrvStackFrame = (comrvStackFrame_t*)((u32_t)pComrvStackFrame & (~D_COMRV_PROFILING_INVOKE_VAL));
```

Comment:
check if we can remove the casting

Reply:
can't be removed; get the following compilation error -
"invalid operands to binary expression"